Make all operations on elements array immutable

GasimGasimzada opened the pull request:
In this PR, I am making all the operations on elements array (not the actual elements within them) immutable. Now, all the functions return values instead of mutating the array directly. Some of the functions to mutate the array directly but currently, those functions accept a copy of the elements array instead of the actual array (`[...elements]`). Those functions are:

- clearSelection
- setSelection
- moveAllLeft / Right
- moveOneLeft / Right

We can also slowly make these functions immutable.

After this PR, I am planning on moving elements array into state as nothing is really dependent on it except for the App component.

dwelle commented:
I've marked `elements` params as `readonly`, where applicable. This adds type checking (which also caught one place where we did mutate the array). I also removed unnecessary copying (e.g. when passing to `clearSelection()` etc. -- since now it's type-checked, we don't need to worry about mistakenly mutating the array).

GasimGasimzada commented:
> I also removed unnecessary copying (e.g. when passing to `clearSelection()` etc. -- since now it's type-checked, we don't need to worry about mistakenly mutating the array).

I added cloning the array on purpose because once we move the array to app state, in order for state updates to work, the original array must differ from new array by reference.

If we do this:

```js
const elements = clearSelection(this.state.elements);
this.setState({ elements });
```

This is not going to update the state because `this.state.elements` and `elements` are the same array. But if we do this:

```js
const elements = clearSelection([...this.state.elements]);
this.setState({ elements });
```

The references change; so, the state will update. We can also do the following when updating the state:

```js
this.setState({ elements: [...elements] });
```

My opinion is that, these function (`clearSelection`, `moveLeftX` etc) must be immutable on their own; however, I didn't want to touch the implementation details. So, I added array copying there.

---

Now that I think about it, it might be more logical to add these arrays inside the functions as if the functions are immutable themselves.

vjeux commented:
Can we hold off landing this after #181 is merged by any chance? Thanks!

giovannigiordano commented:
@GasimGasimzada why this should not work?
```
const elements = clearSelection(this.state.elements);
this.setState({ elements });
```
I expect a copy, not a reference to the same object when I use an immutable function.

GasimGasimzada commented:
@vjeux No problem!

dwelle commented:
@GasimGasimzada AFAIK React's `this.setState` will always re-render (it [doesn't do shallow comparison](https://codesandbox.io/s/pensive-glade-gmgl8)).

Either way, it should be the handler (not the caller), that potentially does the cloning (if necessary), as you said.

GasimGasimzada commented:
@giovannigiordano It's because `clearSelection` is not immutable right now. So, we are imitating immutability. It is the same case in `moveOneLeft`. Instead of rewriting the implementation to be immutable, we are creating a new array and performing operations on the cloned array.

giovannigiordano commented:
@dwelle referencing Immutable.js, it always returns a copy.
@GasimGasimzada for now it's ok.

GasimGasimzada commented:
@dwelle 

> @GasimGasimzada AFAIK React's `this.setState` will always re-render (it [doesn't do shallow comparison](https://codesandbox.io/s/pensive-glade-gmgl8)).

I have been using Hooks for so long that, I have forgotten about it :) I think if there is any interest in migrating to functional components in the future, a safe bet would be using array copies; so that, migration process doesn't break anything.

dwelle commented on src/index.tsx:
```diff
@@ -1077,7 +1077,7 @@ export class App extends React.Component<{}, AppState> {
             if (elementAtPosition && isTextElement(elementAtPosition)) {
               elements = elements.filter(
-                element => element !== elementAtPosition
+                element => element.id !== elementAtPosition.id
```
Not necessary btw

GasimGasimzada replied:
I just did it to not rely on reference comparisons in general. I can revert it back if you want.

dwelle replied:
Just wanted to point it out because it technically borders on voodoo (nothing wrong with reference comparisons where it makes sense --- previous implem of `indexOf` does the same thing), but you can leave it :).

dwelle replied:
Ref comparisons are also slightly faster (you don't need to access a prop and compare strings).

GasimGasimzada replied:
When I think of immutability, I try to avoid reference comparisons at all costs, even if they are for read only accesses like above.

dwelle replied:
Why? Reference comparisons are the bread and butter of immutability. That's how React bails on updates.

GasimGasimzada commented:
Can I use this as a dev branch for the time being and create PRs that merge into this? Then, in the end, we can merge this branch into master.

By changing certain calls to be immutable, I have found new features etc and I want to keep working and pushing without expanding the scope of this branch.

dwelle commented:
@GasimGasimzada I'll review and  merge https://github.com/excalidraw/excalidraw/pull/181 then merge this, so that you don't need to juggling like that. Gimme a bit.

dwelle commented:
Alright, #181 seems to take work before merge-ready. So either continue working and push to this PR, or make new ones against this one (although I'm not sure if it isn't better to just use this PR, otherwise it'll be really hard to keep track of it all).

nanot1m commented on src/index.tsx:
```diff
@@ -215,20 +225,29 @@ export class App extends React.Component<{}, AppState> {
-      elements.forEach(element => {
+      let newElements = [...elements];
+      newElements.forEach(element => {
         element.isSelected = true;
```
One mutation left here

dwelle replied:
Not a mutation of the `elements`, bu of the members. So we're good here.

GasimGasimzada commented:
I was refactoring the code for hit detection and suddenly realized that we can deselect items when clicking shift.

dwelle commented:
> I was refactoring the code for hit detection and suddenly realized that we can deselect items when clicking shift.

Yea, that's part of #272. Btw, I say we merge this.

dwelle commented:
Except, wait, not just yet. I think that `Shift-select` you introduced now isn't working correctly.

GasimGasimzada commented:
You mean merge this branch, right?

GasimGasimzada commented:
> Except, wait, not just yet. I think that `Shift-select` you introduced now isn't working correctly.

What is the problem; so that, I can investigate.

dwelle commented:
Thinking about, can you remove the shift-select commit, merge this PR without it, and make a new PR for it (so that it's on its own. We should also implement https://github.com/excalidraw/excalidraw/issues/273 alongside that).

GasimGasimzada commented:
I have reverted back the changes and fixed a bug there. Should I merge now?

dwelle commented:
Done.

Fausto95 commented on src/index.tsx:
```diff
@@ -1134,15 +1184,15 @@ export class App extends React.Component<{}, AppState> {
 
-      const minX = Math.min(...parsedElements.map(element => element.x));
-      const minY = Math.min(...parsedElements.map(element => element.y));
+      //const minX = Math.min(parsedElements.map(element => element.x));
```
This PR broke the ability of center pasted elements under the cursor

dwelle replied:
Where's that code example from? Can't find any such change from this PR.

dwelle replied:
This is indeed weird change. Didn't catch it.

dwelle replied:
@Fausto95 fix in #291
